python/pycrdt/_transaction.py

```diff
@@ -12,7 +12,38 @@
     from ._doc import Doc
 
 
-class Transaction:
+class BaseTransaction:
+    """
+    Base class for read-write and read-only transactions.
+
+    It allows to persist the origin of the transaction outside of its context.
+    """
+
+    _doc: Doc
+    _origin_hash: int | None
+
+    def __init__(
+        self,
+        doc: Doc,
+        origin: Any = None,
+    ) -> None:
+        self._doc = doc
+        if origin is None:
+            self._origin_hash = None
+        else:
+            self._origin_hash = doc._origins.add(origin)
+
+    def __del__(self) -> None:
+        if getattr(self, "_origin_hash", None) is not None:
+            self._doc._origins.remove(cast(int, self._origin_hash))
```

Comment:
Why using `getattr`, since `_origin_has` is set in `__init__`?

Reply:
I think that it's something to investigate, instead of using `getattr` because we don't know why it doesn't work without. Otherwise that's going to lead to code that we don't understand and don't know how to maintain.